UI: Update NSIS installer script

Fenrirthviti opened the pull request:
### Description
This brings the script in the repo up to date with what is being
used to generate release builds for OBS Studio on Windows.

### Motivation and Context
The installer script is over 15 months old. It's time we pushed what we're actually using up to the repo.

### How Has This Been Tested?
This is the script being used to currently deploy OBS release builds.

### Types of changes
- Housekeeping/Code update

### Checklist:
- [ ] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [X] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [X] My code is not on the master branch.
- [X] The code has been tested.
- [X] All commit messages are properly formatted and commits squashed where appropriate.

### Additional Information
This installer script relies on the tools provided here: https://github.com/notr1ch/OBSInstallerUtils

dodgepong commented on UI/installer/mp-installer.nsi:
```diff
@@ -42,18 +54,14 @@ RequestExecutionLevel admin
 !define MUI_FINISHPAGE_RUN_TEXT "Launch OBS Studio ${SHORTVERSION}"
 !define MUI_FINISHPAGE_RUN_FUNCTION "LaunchOBS"
 
-; GPL is not an EULA, no need to agree to it.
```
This should be kept. The "live" installer probably was not updated to include this.

Fenrirthviti replied:
The current installer still shows the license, the major difference is that the bottom text saying "You are now aware of your rights." was removed. Which part specifically do you want reincluded? The comment in the installer script, or the text itself?

dodgepong replied:
Really it's more the button text that's the biggest issue. The button should say "Next" instead of the default "I agree".

Fenrirthviti replied:
Yes, this is already handled in the new script, on L60.

dodgepong commented on UI/installer/mp-installer.nsi:
```diff
@@ -421,7 +461,7 @@ SectionEnd
 VIAddVersionKey /LANG=${LANG_ENGLISH} "ProductName" "OBS Studio"
 VIAddVersionKey /LANG=${LANG_ENGLISH} "CompanyName" "obsproject.com"
-VIAddVersionKey /LANG=${LANG_ENGLISH} "LegalCopyright" "(c) 2012-2016"
+VIAddVersionKey /LANG=${LANG_ENGLISH} "LegalCopyright" "(c) 2012-2019"
```
Might as well update this too. :P

Fenrirthviti replied:
fixed.

RytoEX replied:
Why not 2012-2020?

dodgepong replied:
You're looking at a comment from an old diff when it still said 2019. The current diff says 2020.

RytoEX replied:
In fairness, normally GitHub would hide a conversation from an old diff, but I had clicked on a link from an email...  In the GitHub UI, I thought you had provided a suggested edit with comment instead of commenting on an existing, older diff.  This is why we don't code review right after waking up.

WizardCM commented on UI/installer/mp-installer.nsi:
```diff
@@ -66,7 +74,14 @@ RequestExecutionLevel admin
+!endif
+
 	${If} ${AtMostWinXP}
 		MessageBox MB_OK|MB_ICONSTOP "Due to extensive use of DirectX 10 features, ${APPNAME} requires Windows Vista SP2 or higher and cannot be installed on this version of Windows."
```
Does OBS even run on Windows Vista these days? Officially we now only support Windows 8 and above, though it does still run on Windows 7. I think it would be a good time to update this check.

Fenrirthviti replied:
In theory, as long as they have SP2 installed. It's definitely not supported, but not sure if it's worth breaking the installer for no good reason.

WizardCM replied:
Hmm, maybe a warning instead, that just says it's "Unsupported"?

Fenrirthviti replied:
What are we saying is unsupported here though? Who are we sending the warning to? Win7 and older? Vista and older? Right now this specific warning is "will OBS even open". I'm open to being convinced, but I'm not sure that putting restrictions or annoying warnings in the installer is the right approach here, or even worth the effort as I don't think we're going to reach a lot of people trying to install OBS on a Vista SP2 machine or older. This is just a CYA as that OS is too old for OBS to even have a hope of running. Doesn't need to be overly complicated. If it was something we saw often, I'd be more receptive but I don't think this is an issue that needs solving.

WizardCM commented on UI/installer/mp-installer.nsi:
```diff
@@ -103,31 +118,38 @@ Function PreReqCheck
+		vs2017true_64:
+			ExecShell "open" "https://obsproject.com/visual-studio-2017-runtimes-64-bit"
+		vs2017false_64:
 		Quit
```
Out of curiosity, would it be possible to change this part of the script to instead show a "Retry" button, rather than immediately closing when the user clicks either button? Taking them out of the flow seems counter-intuitive.

Fenrirthviti replied:
I think it's possible, but in the interest of just getting things up to current standards, going to defer this to a future update to the installer script.

Fenrirthviti commented:
Updated a few things:

- Removed the small installer check at the top, as we will never use that anymore
- Changed the license page to show "License Information" instead of "License Agreement"

This is *mostly* tested by building a version of the installer with many things commented out (browser source, and 32bit files) as I didn't have that stuff built on my system, but I see no reason that it shouldn't work with those file present.

Another important note, this script is very much hardcoded for Jim's specific environment (the /new/core/ folders and such). I am also going to try and make a small page on how to build the installer and make it a little more user friendly for anyone who wants to build an installer for themselves, even though this is mostly going to be used internally by us. It would be good to have the full, repeatable process documented.

Fenrirthviti commented:
Updated again, pushed a few more tweaks to allow for images (currently placeholder) for the installer, and the OBS icon.

![welcome](https://pub.rachni.com/img/OBS-Studio-25.0.8-Full-Installer-x86_2020-04-27_22-33-45.png)
![header](https://pub.rachni.com/img/OBS-Studio-25.0.8-Full-Installer-x86_2020-04-27_22-34-13.png)

Fenrirthviti commented:
A few other potential options we can do, since we can modify the background and text colors as well:

![welcome](https://pub.rachni.com/img/OBS-Studio-25.0.8-Full-Installer-x86_2020-04-27_22-49-20.png)
![license](https://pub.rachni.com/img/OBS-Studio-25.0.8-Full-Installer-x86_2020-04-27_22-49-28.png)

mufunyo commented:
Minor nitpick, but if you're going to remove the word "agreement", it's still mentioned in the text "Press Page Down to see the rest of the agreement".

notr1ch commented:
I can't say I really like the branding of the installer, perhaps because I'm so used to expecting installers to look a certain way it just seems "off".

Fenrirthviti commented:
> Minor nitpick, but if you're going to remove the word "agreement", it's still mentioned in the text "Press Page Down to see the rest of the agreement".

This is already fixed locally:

```!define MUI_LICENSEPAGE_TEXT_TOP "Press Page Down or scroll to see the rest of the license."```

Fenrirthviti commented:
> I can't say I really like the branding of the installer, perhaps because I'm so used to expecting installers to look a certain way it just seems "off".

I'm the opposite way personally. I find it pleasing because of how often it gets neglected. I don't feel particularly strongly about it, and the more I look at it, the less I like the dark theme concept. I do like the banner images and icon though.

WizardCM commented:
NSIS supports the `/S` flag for silent installation (used by apps like [Chocolatey](https://chocolatey.org/packages/obs-studio) and the newly announced [Windows Package Manager](https://github.com/microsoft/winget-pkgs)).

The problem is that our Visual Studio runtime check (officially) only occurs in the UI - so if you run our installer with the `/S` flag, the check is ignored and OBS installs. On first launch, you end up seeing an error (which doesn't describe the problem well).

We could use [IfSilent](https://nsis.sourceforge.io/Reference/IfSilent) to detect this and do something different, but as we don't ship the runtimes directly and normally open a webpage with a link to the runtimes, I'm not sure what the best course of action is here.

RytoEX commented:
@WizardCM 
I wouldn't let that hold this up further though.  Honestly, I think we just should commit the updated installer script as-is for what we're actually using, and then do further updates and improvements as separate PRs/commits rather than holding this up.

WizardCM commented:
@RytoEX i agree, just wanted to document it somewhere for later.

Fenrirthviti commented:
For reference, it will be possible with CI to pass the version numbers by using variables during the installation. Relevant references:

https://github.com/Fenrirthviti/obs-virtual-cam/blob/master/installer/installer.nsi#L6
https://github.com/Fenrirthviti/obs-virtual-cam/blob/master/.github/workflows/main.yml#L160

I have given a final check over this, and it looks like it is good to go.

Fenrirthviti commented on UI/installer/mp-installer.nsi:
```diff
@@ -258,102 +318,71 @@ Section "OBS Studio" SecCore
+!endif
 
-	SetOutPath "$INSTDIR\bin\32bit"
+	CreateShortCut "$SMPROGRAMS\OBS Studio\Uninstall.lnk" "$INSTDIR\uninstall.exe"
```
For a change like that, I would prefer that we wait until the current version is merged. I'm already regretting making the small changes I did as it muddies the history even further. I agree though, but this should be part of a different PR.

RytoEX replied:
I agree wholeheartedly, as before.  Let's get an up-to-date installer merged, then work on further incremental improvements.

Fenrirthviti commented on UI/installer/mp-installer.nsi:
```diff
@@ -258,102 +318,71 @@ Section "OBS Studio" SecCore
+	CreateShortCut "$SMPROGRAMS\OBS Studio\OBS Studio (64bit).lnk" "$INSTDIR\bin\64bit\obs64.exe"
+!else
 	SetOutPath "$INSTDIR\bin\32bit"
 	CreateDirectory "$SMPROGRAMS\OBS Studio"
```
Took me a bit to see what you were referring to, but I am not currently sure why there is a create directory outside the ifdef. NSIS does a lot of things based on the bitness that it's compiled for, so there might be some unintuitive behavior where this is the correct way of handling it.

RytoEX replied:
I have some notes of things I want to address in the installer regarding the 32-bit and 64-bit installer commands, but I'm going to worry about those later.
